specs/SharedBuffer.md

```diff
@@ -0,0 +1,282 @@
+  /// 
+  HRESULT ShareBufferToScript(
+    [in] ICoreWebView2StagingSharedBuffer* sharedBuffer,
+    [in] BOOL isReadOnlyToScript,
+    [in] LPCWSTR additionalDataAsJson);
+}
+
+interface ICoreWebView2StagingFrame2 : IUnknown {
+  /// Share a shared buffer object with script of the iframe in the WebView.
+  /// The script will receive a `SharedBufferReceived` event from chrome.webview.
+  /// The event arg for that event will have the following properties:
+  ///   `sharedBuffer`: an ArrayBuffer object with the backing content from the shared buffer.
+  ///   `data`: an object representing `additionalDataAsJson`. This property will be `undefined`
+  ///           if `additionalDataAsJson` is nullptr or empty string.
+  ///   `source`: with a value set as `chrome.webview` object.
+  /// If `isReadOnlyToScript` is true, the script will only have read access to the buffer.
+  /// If the script tries to modify the content in a read only buffer, it will cause an access violation
+  /// in WebView renderer process and crash the renderer process.
+  /// If the shared buffer is already closed, the API will fail with HRESULT_FROM_WIN32(ERROR_INVALID_STATE).
+  /// 
+  /// The script code should call `chrome.webview.releaseSharedBuffer` with
+  /// the shared buffer as the parameter to release underlying resources as soon
+  /// as it does not need access the shared buffer any more.
+  ///
+  /// Sharing a buffer to script has security risk. You should only share buffer with trusted site.
+  /// 
+  HRESULT ShareBufferToScript(
+    [in] ICoreWebView2StagingSharedBuffer* sharedBuffer,
+    [in] BOOL isReadOnlyToScript,
+    [in] LPCWSTR additionalDataAsJson);
+}
+
+```
+
+## .NET
+```c#
+namespace Microsoft.Web.WebView2.Core
+{
+
+    class CoreWebView2Environment
+    {
+        public CoreWebView2SharedBuffer CreateSharedBuffer(ulong Size);
+    }
+
+    class CoreWebView2SharedBuffer
+    {
+        public ulong Size { get; };
+        public IntPtr Buffer { get; };
+        public IntPtr Handle { get; };
+        public Stream GetAccessStream();
+        void Close();
+    }
+    
+    runtimeclass CoreWebView2
+    {
+        public void ShareBufferToScript(CoreWebView2SharedBuffer sharedBuffer, bool isReadOnlyToScript, string additionalDataAsJson);
+    }
+    
+    class CoreWebView2Frame
+    {
+        public void ShareBufferToScript(CoreWebView2SharedBuffer sharedBuffer, bool isReadOnlyToScript, string additionalDataAsJson);
+    }
+}
+
+```
+## WinRT
+```c#
+namespace Microsoft.Web.WebView2.Core
+{
+
+    runtimeclass CoreWebView2Environment
+    {
+        [interface_name("Microsoft.Web.WebView2.Core.ICoreWebView2StagingEnvironment2")]
+        {
+            CoreWebView2SharedBuffer CreateSharedBuffer(UInt64 Size);
+        }
+    }
+
+    runtimeclass CoreWebView2SharedBuffer
+    {
+        UInt64 Size { get; };
+        Windows.Storage.Streams.IRandomAccessStream GetAccessStream();
```

Comment:
For all 3 of the SharedBuffer classes, should they just implement their stream interface directly rather than having a separate method and object?

Reply:
It might be better to have a separate object for accessing the buffer. A buffer is a piece of memory. An IStream is created for accessing the memory, it is not the same as the buffer itself.
It is actually easier for applications to create a stream per access, read/write the buffer and then dispose the stream. If we implement the stream on the buffer object itself, the app has to be careful about not disposing the buffer itself accidentally.

Reply:
Yeah I came to that conclusion last night too. Stream has its own state about where it is in the buffer and such. Makes sense to have different stream objects that can point to different places in the buffer.